File: 2026-protocol/nextcloud-file-browser/identity.go
package main

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"io"
	"log"
	"math/big"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gorilla/websocket"
)

// PendingAuth stores authentication state for a connection
type PendingAuth struct {
	Nonce     [32]byte
	Timestamp int64
	ActionID  uint64
}

// ConnectionState tracks the state of a WebSocket connection
type ConnectionState struct {
	Conn         *websocket.Conn
	Identity     string // "username@server"
	NextcloudURL string // "https://server"
	Username     string
	AppPassword  string
	State        string // awaiting_identity, awaiting_auth, browsing
	PendingAuth  *PendingAuth
	LoginPoll    *LoginPollState
}

// LoginPollState tracks Login Flow v2 polling
type LoginPollState struct {
	PollEndpoint string
	PollToken    string
	StopChan     chan struct{}
}

// PublicKeyData contains the public key and embedded identity
type PublicKeyData struct {
	Identity string
	Key      *ecdsa.PublicKey
}

func sendIdentificationRequest(conn *websocket.Conn) (*PendingAuth, error) {
	// Generate nonce
	var nonce [32]byte
	if _, err := rand.Read(nonce[:]); err != nil {
		return nil, fmt.Errorf("failed to generate nonce: %w", err)
	}

	timestamp := time.Now().Unix()
	actionID := uint64(1)

	pending := &PendingAuth{
		Nonce:     nonce,
		Timestamp: timestamp,
		ActionID:  actionID,
	}

	// Build message: type (1) + action_id (8) + nonce (32) + timestamp (8) + reason (string)
	reason := "Nextcloud file browser needs to verify your identity"
	msg := make([]byte, 1+8+32+8+len(reason))
	msg[0] = msgServerRequestIdentification
	binary.BigEndian.PutUint64(msg[1:9], actionID)
	copy(msg[9:41], nonce[:])
	binary.BigEndian.PutUint64(msg[41:49], uint64(timestamp))
	copy(msg[49:], reason)

	log.Printf("Sending identification request (nonce: %x...)", nonce[:8])
	return pending, conn.WriteMessage(websocket.BinaryMessage, msg)
}

func handleIdentificationResponse(state *ConnectionState, msg []byte) {
	// Parse: type (1) + action_id (8) + identity_url (null-terminated) + signature (64)
	if len(msg) < 1+8+1+64 {
		log.Printf("Identification response too short")
		return
	}

	actionID := binary.BigEndian.Uint64(msg[1:9])

	// Find null terminator for identity URL
	urlEnd := -1
	for i := 9; i < len(msg)-64; i++ {
		if msg[i] == 0 {
			urlEnd = i
			break
		}
	}
	if urlEnd == -1 {
		log.Printf("No null terminator found in identity URL")
		return
	}

	identityURL := string(msg[9:urlEnd])
	signature := msg[urlEnd+1:]

	if len(signature) != 64 {
		log.Printf("Invalid signature length: %d (expected 64)", len(signature))
		return
	}

	log.Printf("Received identification response: %s", identityURL)

	pending := state.PendingAuth
	if pending == nil {
		log.Printf("No pending auth for connection")
		return
	}
	state.PendingAuth = nil

	if actionID != pending.ActionID {
		log.Printf("Action ID mismatch: got %d, expected %d", actionID, pending.ActionID)
		return
	}

	// Check timestamp is within ±5 minutes
	now := time.Now().Unix()
	if pending.Timestamp < now-300 || pending.Timestamp > now+300 {
		log.Printf("Timestamp out of range: %d (now: %d)", pending.Timestamp, now)
		return
	}

	// Parse the identity URL to get the hosting server
	parsedURL, err := url.Parse(identityURL)
	if err != nil {
		log.Printf("Failed to parse identity URL: %v", err)
		return
	}
	hostingServer := parsedURL.Host

	// Fetch public key from identity URL
	pubKeyData, err := fetchPublicKey(identityURL)
	if err != nil {
		log.Printf("Failed to fetch public key: %v", err)
		return
	}

	log.Printf("Fetched public key with identity: %s", pubKeyData.Identity)

	// Parse identity from public key file (format: username@server)
	identity := pubKeyData.Identity
	parts := strings.SplitN(identity, "@", 2)
	if len(parts) != 2 {
		log.Printf("Invalid identity format: %s", identity)
		return
	}
	username := parts[0]
	server := parts[1]

	// Verify that the claimed server matches the hosting server
	if !strings.EqualFold(server, hostingServer) {
		log.Printf("Identity server mismatch: claimed %s but hosted on %s", server, hostingServer)
		return
	}

	// Verify signature
	signedData := make([]byte, 40)
	copy(signedData[:32], pending.Nonce[:])
	binary.BigEndian.PutUint64(signedData[32:], uint64(pending.Timestamp))

	hash := sha256.Sum256(signedData)

	r := new(big.Int).SetBytes(signature[:32])
	s := new(big.Int).SetBytes(signature[32:64])

	if !ecdsa.Verify(pubKeyData.Key, hash[:], r, s) {
		log.Printf("Signature verification failed")
		return
	}

	log.Printf("Signature verified successfully!")
	log.Printf("Identified user: %s@%s", username, server)

	// Store identity info
	state.Identity = identity
	state.Username = username
	state.NextcloudURL = fmt.Sprintf("https://%s", server)

	// Check if we have stored credentials for this identity
	credStoreMu.Lock()
	cred, exists := credStore.Get(identity)
	credStoreMu.Unlock()

	if exists {
		log.Printf("Found stored credentials for %s", identity)
		state.Username = cred.Username
		state.AppPassword = cred.AppPassword
		state.NextcloudURL = cred.NextcloudURL
		state.State = stateBrowsing
		// Send root directory listing
		sendDirectoryListing(state, "/")
	} else {
		log.Printf("No credentials for %s, starting Nextcloud auth flow", identity)
		startNextcloudAuth(state)
	}
}

func fetchPublicKey(identityURL string) (*PublicKeyData, error) {
	// Add /download to get the raw file
	downloadURL := identityURL + "/download"

	resp, err := http.Get(downloadURL)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	// Read public key file
	// Format: identity_string (null-terminated) + public key (65 bytes: 0x04 || X || Y)
	keyData, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read body: %w", err)
	}

	// Find null terminator to split identity from key
	nullIdx := bytes.IndexByte(keyData, 0)
	if nullIdx == -1 {
		return nil, fmt.Errorf("no identity found in public key file")
	}

	identity := string(keyData[:nullIdx])
	keyBytes := keyData[nullIdx+1:]

	// Parse uncompressed point format (0x04 || X || Y)
	if len(keyBytes) != 65 || keyBytes[0] != 0x04 {
		return nil, fmt.Errorf("invalid public key format (len=%d)", len(keyBytes))
	}

	x := new(big.Int).SetBytes(keyBytes[1:33])
	y := new(big.Int).SetBytes(keyBytes[33:65])

	pubKey := &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     x,
		Y:     y,
	}

	return &PublicKeyData{
		Identity: identity,
		Key:      pubKey,
	}, nil
}

File: 2026-protocol/nextcloud-connector/src/identity.rs
//! Gradesta identity verification

use anyhow::{anyhow, Context, Result};
use p256::ecdsa::{signature::Verifier, Signature, VerifyingKey};
use std::time::{SystemTime, UNIX_EPOCH};
use url::Url;

use crate::protocol;

/// Pending authentication state
#[derive(Debug, Clone)]
pub struct PendingAuth {
    pub nonce: [u8; 32],
    pub timestamp: i64,
    pub action_id: u64,
}

/// Public key data fetched from identity URL
#[derive(Debug)]
pub struct PublicKeyData {
    /// Embedded claim (username@server) - unverified, only used for Nextcloud URL derivation
    pub embedded_claim: String,
    pub key: VerifyingKey,
}

impl PendingAuth {
    /// Create new pending auth with random nonce
    pub fn new(action_id: u64) -> Self {
        let mut nonce = [0u8; 32];
        rand::RngCore::fill_bytes(&mut rand::thread_rng(), &mut nonce);
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        Self {
            nonce,
            timestamp,
            action_id,
        }
    }

    /// Encode the request identification message
    pub fn encode_request(&self, reason: &str) -> Vec<u8> {
        protocol::encode_request_identification(self.action_id, &self.nonce, self.timestamp, reason)
    }
}

/// Verified identity result
#[derive(Debug)]
pub struct VerifiedIdentity {
    /// The identity URL (the real identity - where the public key is hosted)
    pub identity_url: String,
    /// The embedded claim from the public key (username@server)
    pub embedded_claim: String,
}

/// Verify an identification response
pub async fn verify_identification(
    pending: &PendingAuth,
    action_id: u64,
    identity_url: &str,
    signature: &[u8],
) -> Result<VerifiedIdentity> {
    // Check action ID matches
    if action_id != pending.action_id {
        return Err(anyhow!(
            "Action ID mismatch: got {}, expected {}",
            action_id,
            pending.action_id
        ));
    }

    // Check timestamp is within ±5 minutes
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;
    if pending.timestamp < now - 300 || pending.timestamp > now + 300 {
        return Err(anyhow!(
            "Timestamp out of range: {} (now: {})",
            pending.timestamp,
            now
        ));
    }

    // Parse identity URL to get hosting server
    let parsed_url = Url::parse(identity_url).context("Failed to parse identity URL")?;
    let hosting_server = parsed_url
        .host_str()
        .ok_or_else(|| anyhow!("No host in identity URL"))?;

    // Fetch public key
    let pub_key_data = fetch_public_key(identity_url).await?;

    log::info!("Fetched public key with embedded claim: {}", pub_key_data.embedded_claim);

    // Parse embedded identity claim (format: username@server)
    // This is just a claim - the real identity is the URL where the key is hosted
    let parts: Vec<&str> = pub_key_data.embedded_claim.splitn(2, '@').collect();
    if parts.len() != 2 {
        return Err(anyhow!("Invalid embedded identity format: {}", pub_key_data.embedded_claim));
    }
    let claimed_username = parts[0];
    let claimed_server = parts[1];

    // Verify claimed server matches hosting server
    // This ensures someone can't claim to be user@example.com while hosting on evil.com
    if !claimed_server.eq_ignore_ascii_case(hosting_server) {
        return Err(anyhow!(
            "Identity server mismatch: claimed {} but hosted on {}",
            claimed_server,
            hosting_server
        ));
    }

    // Verify signature
    // Signed data: nonce (32) || timestamp (8)
    // Note: p256's Signer::sign hashes internally, so we verify against the raw data
    let mut signed_data = Vec::with_capacity(40);
    signed_data.extend_from_slice(&pending.nonce);
    signed_data.extend_from_slice(&(pending.timestamp as u64).to_be_bytes());

    // Parse signature (r || s, 32 bytes each)
    if signature.len() != 64 {
        return Err(anyhow!("Invalid signature length: {} (expected 64)", signature.len()));
    }

    let sig = Signature::from_slice(signature).context("Failed to parse signature")?;

    // Verify against raw data - the Verifier trait will hash it internally
    pub_key_data
        .key
        .verify(&signed_data, &sig)
        .map_err(|e| anyhow!("Signature verification failed: {:?}", e))?;

    log::info!("Signature verified successfully!");
    log::info!("Identity URL (real identity): {}", identity_url);
    log::info!("Embedded claim (verified): {}", pub_key_data.embedded_claim);

    // Return both the URL and the embedded claim
    Ok(VerifiedIdentity {
        identity_url: identity_url.to_string(),
        embedded_claim: pub_key_data.embedded_claim,
    })
}

/// Fetch public key from identity URL
async fn fetch_public_key(identity_url: &str) -> Result<PublicKeyData> {
    let download_url = format!("{}/download", identity_url);

    let response = reqwest::get(&download_url)
        .await
        .context("Failed to fetch public key")?;

    if !response.status().is_success() {
        return Err(anyhow!("HTTP {}", response.status()));
    }

    let key_data = response.bytes().await.context("Failed to read response")?;

    // Parse: identity_string (null-terminated) + public key (65 bytes: 0x04 || X || Y)
    let null_idx = key_data
        .iter()
        .position(|&b| b == 0)
        .ok_or_else(|| anyhow!("No identity found in public key file"))?;

    let identity = String::from_utf8(key_data[..null_idx].to_vec())
        .context("Invalid identity string")?;
    let key_bytes = &key_data[null_idx + 1..];

    // Parse uncompressed point format (0x04 || X || Y)
    if key_bytes.len() != 65 || key_bytes[0] != 0x04 {
        return Err(anyhow!(
            "Invalid public key format (len={})",
            key_bytes.len()
        ));
    }

    let verifying_key = VerifyingKey::from_sec1_bytes(key_bytes)
        .context("Failed to parse public key")?;

    Ok(PublicKeyData {
        embedded_claim: identity,
        key: verifying_key,
    })
}
